File: scala/src/main/scala/Sorting.scala
import scala.util.Random

/**
 * Quicksort implementation for performance benchmarking.
 * Tests with 100,000 random integers.
 */
object Sorting {
  /**
   * Quicksort algorithm implementation.
   */
  def quicksort(arr: List[Int]): List[Int] = {
    if (arr.length <= 1) arr
    else {
      val pivot = arr(arr.length / 2)
      val (left, rest) = arr.partition(_ < pivot)
      val (middle, right) = rest.partition(_ == pivot)
      quicksort(left) ::: middle ::: quicksort(right)
    }
  }

  /**
   * Run sorting benchmark with 100,000 random integers.
   */
  def runBenchmark(): Unit = {
    val random = new Random()
    val arr = List.fill(100000)(random.nextInt(1000000))

    val startTime = System.nanoTime()
    val sortedArr = quicksort(arr)
    val endTime = System.nanoTime()

    val executionTime = (endTime - startTime) / 1000000.0 // Convert to milliseconds

    // Verify sorting correctness
    val isSorted = sortedArr.sliding(2).forall {
      case List(a, b) => a <= b
      case _ => true
    }

    println("\n=== Sorting (Quicksort) ===")
    println(s"Array size: ${arr.length}")
    println(f"Execution time: $executionTime%.2f ms")
    println(s"Correctly sorted: $isSorted")
  }

  def main(args: Array[String]): Unit = {
    runBenchmark()
  }
}

File: csharp/Sorting.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace StackWizard
{
    /// <summary>
    /// Quicksort implementation for performance benchmarking.
    /// Tests with 100,000 random integers.
    /// </summary>
    public class Sorting
    {
        /// <summary>
        /// Quicksort algorithm implementation.
        /// </summary>
        public static List<int> Quicksort(List<int> arr)
        {
            if (arr.Count <= 1)
                return arr;

            int pivot = arr[arr.Count / 2];
            var left = arr.Where(x => x < pivot).ToList();
            var middle = arr.Where(x => x == pivot).ToList();
            var right = arr.Where(x => x > pivot).ToList();

            var result = new List<int>();
            result.AddRange(Quicksort(left));
            result.AddRange(middle);
            result.AddRange(Quicksort(right));
            return result;
        }

        /// <summary>
        /// Run sorting benchmark with 100,000 random integers.
        /// </summary>
        public static void RunBenchmark()
        {
            var random = new Random();
            var arr = Enumerable.Range(0, 100000)
                                .Select(_ => random.Next(0, 1000000))
                                .ToList();

            var watch = System.Diagnostics.Stopwatch.StartNew();
            var sortedArr = Quicksort(arr);
            watch.Stop();

            double executionTime = watch.Elapsed.TotalMilliseconds;

            // Verify sorting correctness
            bool isSorted = Enumerable.Range(0, sortedArr.Count - 1)
                                     .All(i => sortedArr[i] <= sortedArr[i + 1]);

            Console.WriteLine("\n=== Sorting (Quicksort) ===");
            Console.WriteLine($"Array size: {arr.Count}");
            Console.WriteLine($"Execution time: {executionTime:F2} ms");
            Console.WriteLine($"Correctly sorted: {isSorted}");
        }
    }
}
